---
title: "Melanoma scRNAseq feature table build"
author: "Marco Tello"
date: "`r Sys.Date()`"
output: html_document
runtime: shiny
---

```{r setup, include=FALSE}
library(data.table)
library(Rtsne)
library(ggplot2)
library(scales)
library(RColorBrewer)
library(dplyr)
library(irlba)
knitr::opts_chunk$set(echo = FALSE)
```


```{r}
exp_dt <- fread("data/GSE72056_exp_df.tsv")
cell_dt <- fread("data/GSE72056_cell_metadata.tsv")
cell_dt <- cell_dt[malignant != "Unresolved"]
cell_dt <- cell_dt[cell != "Unresolved"]
cell_dt[, tumor_id := factor(tumor_id)]
cell_dt[, malignant := NULL]
cell_dt[, cell := factor(cell, 
                         levels = c("T", "B", "Macrophage", 
                                    "Endothelial", "CAF",
                                    "NaturalKiller"))]
donor_dt <- fread("data/GSE72056_donor_metadata.tsv")

tumors <- as.character(unique(cell_dt$tumor_id))
celltypes <- unique(cell_dt$cell)

```



# Goal of the current Shiny document: 

The current document is aimed to provide an proof of concept for an interactive environment to visualize the single cell RNA-seq data from the study "Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq" by Tirosh, and collaborators 2016.


# Dataset description. 

The single cell dataset is composed of 4,645 cells derived from 18 different donors. Each cell was further classified into different cell types based on the expression of marker genes and into malignant status based on estimated copy number variation (CNV). Due to the file size and limitations of an online Shiny app, I only included the non-malignant cells from two tumors.


```{r}
summary(cell_dt[, .SD, .SDcols = c("tumor_id", "cell")], maxsum = 20, na.strings = "")

```


# General distribution of metadata:

The first step is to visualize the distribution of cells by each category of malignant or cell types.

```{r}

color_df <- data.table(
  Variable = factor(x = levels(cell_dt$cell), 
                    levels = levels(cell_dt$cell)),
  Color = factor(brewer.pal(6, "Set2"))
)
data_plot <- reactive({
  plot_dt <- cell_dt[, .SD, .SDcols = c("cell_id", "tumor_id", "cell")]
  plot_dt <- merge.data.table(x = plot_dt, y =  color_df, 
                              by.x = "cell", by.y = "Variable", 
                              all.x = TRUE)
})


renderPlot({
  plot_dt <- data_plot()
  fill_dt <- data.table(Color = levels(droplevels(plot_dt$Color)), 
                        Variable = levels(unlist(plot_dt[, .SD, .SDcols = "cell"])))
  ggplot(plot_dt, aes(x = tumor_id, fill = cell)) +
    geom_bar(position = "fill") +
    labs(title = element_blank(), y = "Fraction", x = "Tumor ID") +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = fill_dt$Color, labels = fill_dt$Variable) +
    theme_classic() 
})

renderPlot({
  plot_dt <- data_plot()
  fill_dt <- data.table(Color = levels(droplevels(plot_dt$Color)), 
                        Variable = levels(unlist(plot_dt[, .SD, .SDcols = "cell"])))
  ggplot(plot_dt, aes(x = tumor_id, fill = cell)) +
    geom_bar(position = position_dodge()) +
    labs(title = element_blank(), y = "Number of cells", x = "Tumor ID") +
    scale_y_continuous(trans = "log10", labels = scales::comma) +
    scale_fill_manual(values = fill_dt$Color, labels = fill_dt$Variable) +
    theme_classic() 
})
```


# tSNE visualization and cell filtering

The original study from Tirosh and collaborators 2016 performed tSNE plots
to visualize the organization of cells into clusters based on their transcriptomic profiles. 
  
  
The following code reproduces the tSNE plot for figure 1C, however it is possible to select different options to also reproduce figure 1D or visualize other combinations of cells. 

To calculate the tSNE plot based on the selected cells jsut click on "Generate tSNE plot". Running time could take up to one minute, depending on the amount of cells included (all boxes checked).    
 
Cells are automatically colored by estimated cell type.

Please note that the subset of cells used to generate the last tSNE will be used in the following sections. For instance, if you want to visualize variability only between T and B cells you must select the corresponding boxes and then generate the corresponding tSNE plot. This decision is aimed to maintain a consistent data subset during the rest of visualizations.

```{r}

color_df <- data.table(
  Variable = factor(x = c(levels(cell_dt$cell), levels(cell_dt$tumor_id)), 
                    levels = c(levels(cell_dt$cell), levels(cell_dt$tumor_id))),
  Color = factor(c(brewer.pal(6, "Set2"), brewer.pal(3, "Dark2")[1:2]))
)


inputPanel(
  checkboxGroupInput("celltypes", label = "Cell types:",
                     choices = celltypes, selected = celltypes),
  checkboxGroupInput("tumor_id", label = "Tumors to include:",
                     choices = sort(tumors), selected =c("79","80","84")),
  actionButton("updatePlot", "Generate tSNE plot")
  
)
inputPanel(
  selectInput("colorby", label = "Color cells by:",
              choices = c("Tumor"= "tumor_id", 
                          "Cell type" = "cell"), 
              selected = "tumor_id")
)
updatePlotClicked <- reactiveVal(FALSE)  # Track if the "Update Plot" button is clicked
tsnePlot <- reactiveVal(NULL)  # Store the tsne_plot

observeEvent(input$updatePlot, {
  updatePlotClicked(TRUE)
  tsnePlot(updateTsnePlot())  # Update the stored tsne_plot
  updatePlotClicked(FALSE)  # Reset the value to avoid further automatic re-executions
})

updateTsnePlot <- function() {
  req(updatePlotClicked())  # Only proceed if the updatePlot button is clicked
  cell2plot <- cell_dt[cell %in% input$celltypes][tumor_id %in% input$tumor_id, cell_id]
  
  dt_tsne <- t(as.matrix(exp_dt[, .SD, .SDcols = cell2plot]))
  tsne_plot <- Rtsne(dt_tsne, partial_pca = TRUE, verbose = TRUE, num_threads = 1)
  tsne_plot <- data.table(cell_id = rownames(dt_tsne),
                          x = tsne_plot$Y[,1],
                          y = tsne_plot$Y[,2])
  tsne_plot <- merge.data.table(tsne_plot, cell_dt, by = "cell_id")
  return(tsne_plot)
}

dataInput <- reactive({
  req(!is.null(tsnePlot())) # Ensure tsnePlot is not NULL
  tsnePlot()
})

renderPlot({
  plot_dt <- dataInput()
  plot_dt <- merge.data.table(x = plot_dt, y =  color_df, 
                              by.x = input$colorby, by.y = "Variable", 
                              all.x = TRUE)
  fill_dt <- data.table(Color = levels(droplevels(plot_dt$Color)), 
                        Variable = levels(droplevels(unlist(plot_dt[, .SD, .SDcols = input$colorby]))))
  ggplot(plot_dt) +
    geom_point(aes(x = x, y = y, color = !!sym(input$colorby))) +
    labs(x = "tSNE1", y = "tSNE2")+
    scale_color_manual(values = fill_dt$Color, labels = fill_dt$Variable) +
    theme_classic()
})

```



# Gene expression variability

A common approach to reduce dimensionality is to focus on genes whose expression is highly variable across the cells in the dataset. The following section allows you to visualize the expression variability using median absolute deviation (MAD) scores. 


```{r, error = FALSE}
inputPanel(
  selectInput("n_breaks", label = "Number of bins:",
              choices = c(20, 50, 100), selected = 20)
)

hist_dt_reactive <- reactiveVal(NULL)

sliderInput("var_threshold", label = "Select variability threshold:",
             min = 0, max = 5, value = 0)

observeEvent(input$updatePlot, {
  hist_dt_reactive(updateHist())
  all_var <- hist_dt_reactive()$exp_var
  updateSliderInput(session, "var_threshold", 
                    min = 0,
                    max = round(max(all_var), 2),
                    value = 0, 
                    step = 0.001)
  output$histogram <- renderPlot({
  plot_dt <- hist_dt_reactive()
  ggplot(plot_dt, aes(x = exp_var)) +
    geom_histogram(bins = input$n_breaks) +
    scale_y_continuous(trans = "log10", labels = scales::comma) +
    geom_vline(xintercept = input$var_threshold, 
               linetype = "dashed", color = "red") +
    labs(y = "Number of genes", x = "MAD scores", 
         title = paste("Current threshold will conserve ~", 
                       round((1-(ecdf(all_var)(input$var_threshold)))*100), 
                       "% of genes", 
                       sep = "")) +
    theme_classic()
})
})

updateHist <- function() {
  cell2plot <- cell_dt[cell %in% input$celltypes][tumor_id %in% input$tumor_id, cell_id]
  exp_mat <- as.matrix(exp_dt[, .SD, .SDcols = cell2plot])
  gene_var <- apply(exp_mat, MARGIN = 1, mad, na.rm = TRUE)
  gene_var <- round(gene_var, 2) + 0.001
  hist_dt <- data.table(ENSG = exp_dt$ENSG, 
                        exp_var = gene_var)
  return(hist_dt)
}

plotOutput("histogram")

```


